packages/blob-storage/indexer/indexer.go

```diff
@@ -0,0 +1,388 @@
+	Data []struct {
+		Index            string `json:"index"`
+		Blob             string `json:"blob"`
+		KzgCommitment    string `json:"kzg_commitment"`
+		KzgCommitmentHex []byte `json:"-"`
+	} `json:"data"`
+}
+
+// Indexer struct holds the configuration and state for the Ethereum chain listener.
+type Indexer struct {
+	beaconURL   string
+	ethClient   *ethclient.Client
+	startHeight *uint64
+	taikoL1     *taikol1.TaikoL1
+	db          *mongodb.MongoDBClient
+	cfg         *Config
+	wg          *sync.WaitGroup
+	ctx         context.Context
+}
+
+func (i *Indexer) InitFromCli(ctx context.Context, c *cli.Context) error {
+	cfg, err := NewConfigFromCliContext(c)
+	if err != nil {
+		return err
+	}
+
+	return InitFromConfig(ctx, i, cfg)
+}
+
+// InitFromConfig inits a new Indexer from a provided Config struct
+func InitFromConfig(ctx context.Context, i *Indexer, cfg *Config) (err error) {
+	client, err := ethclient.Dial(cfg.RPCURL)
+	if err != nil {
+		return err
+	}
+
+	taikoL1, err := taikol1.NewTaikoL1(cfg.ContractAddress, client)
+	if err != nil {
+		return err
+	}
+
+	db, err := mongodb.NewMongoDBClient(mongodb.MongoDBConfig{
+		Host:     cfg.DBHost,
+		Port:     cfg.DBPort,
+		Username: cfg.DBUsername,
+		Password: cfg.DBPassword,
+		Database: cfg.DBDatabase,
+	})
+	if err != nil {
+		return err
+	}
+
+	i.ethClient = client
+	i.beaconURL = cfg.BeaconURL
+	i.taikoL1 = taikoL1
+	i.startHeight = cfg.StartingBlockID
+	i.db = db
+	i.wg = &sync.WaitGroup{}
+	i.ctx = ctx
+	i.cfg = cfg
+
+	return nil
+}
+
+func (i *Indexer) Start() error {
+	i.wg.Add(1)
+
+	var processingBlockHeight uint64
+
+	if i.startHeight != nil {
+		processingBlockHeight = *i.startHeight
+	} else {
+		n, err := i.getLatestBlockID()
+		if err != nil {
+			return err
+		}
+
+		processingBlockHeight = n
+	}
+
+	// get the latest header
+	header, err := i.ethClient.HeaderByNumber(i.ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// the end block is the latest header.
+	endBlockID := header.Number.Uint64()
+
+	var defaultBlockBatchSize uint64 = 50
+
+	slog.Info("fetching batch block events",
+		"startHeight", processingBlockHeight,
+		"endblock", endBlockID,
+		"batchsize", defaultBlockBatchSize,
+	)
+
+	for j := processingBlockHeight; j < endBlockID; j += defaultBlockBatchSize {
+		end := processingBlockHeight + uint64(defaultBlockBatchSize)
+		// if the end of the batch is greater than the latest block number, set end
+		// to the latest block number
+		if end > endBlockID {
+			end = endBlockID
+		}
+
+		// filter exclusive of the end block.
+		// we use "end" as the next starting point of the batch, and
+		// process up to end - 1 for this batch.
+		filterEnd := end - 1
+
+		slog.Info("block batch", "start", j, "end", filterEnd)
+
+		opts := &bind.FilterOpts{
+			Start:   processingBlockHeight,
+			End:     &filterEnd,
+			Context: i.ctx,
+		}
+
+		group, groupCtx := errgroup.WithContext(i.ctx)
+
+		events, err := i.taikoL1.FilterBlockProposed(opts, nil, nil)
```

Comment:
Not sure if it is simpler for this repo, shall we just `import "github.com/taikoxyz/taiko-client/pkg/chain_iterator/event_iterator"` package here, and reuse `BlockProposedIterator`?

Reply:
I'll take a look tomorrow, perhaps we wait until Taiko client is merged into this codebase otherwise the imports will break when we delete the Taiko client repo